packages/viewer/src/components/scene-tree-table/scene-tree-table.tsx

```diff
@@ -0,0 +1,511 @@
+
+  public async componentWillRender(): Promise<void> {
+    if (this.controller?.isConnected) {
+      await this.controller.updateActiveRowRange(
+        this.stateMap.startIndex,
+        this.stateMap.endIndex
+      );
+    }
+
+    this.computeViewportRows();
+  }
+
+  public componentDidRender(): void {
+    this.layoutColumns();
+  }
+
+  public disconnectedCallback(): void {
+    this.columnElements.forEach((c) => {
+      c.removeEventListener('hovered', this.handleCellHover as EventListener);
+    });
+    this.tableElement?.removeEventListener('scroll', this.handleScrollChanged);
+  }
+
+  @Method()
+  public async getViewportStartIndex(): Promise<number> {
+    return this.stateMap.startIndex;
+  }
+
+  @Method()
+  public async getViewportEndIndex(): Promise<number> {
+    return this.stateMap.endIndex;
+  }
+
+  public render(): h.JSX.IntrinsicElements {
+    return (
+      <Host>
+        <div
+          class="header"
+          ref={(ref) => (this.headerElement = ref)}
+          style={{
+            gridTemplateColumns: this.columnGridLayout,
+          }}
+        />
+        <div
+          class="table"
+          ref={(ref) => (this.tableElement = ref)}
+          style={{
+            gridTemplateColumns: this.columnGridLayout,
+          }}
+          onScroll={this.handleScrollChanged}
+        >
+          <slot onSlotchange={this.bindHeaderData} />
+        </div>
+      </Host>
+    );
+  }
+
+  private computeViewportRows(): void {
+    const viewportHeight = this.getLayoutHeight();
+    if (viewportHeight != null) {
+      const viewportCount = Math.ceil(viewportHeight / this.rowHeight);
+
+      const viewportStartIndex = Math.floor(this.scrollOffset / this.rowHeight);
+      const viewportEndIndex = viewportStartIndex + viewportCount;
+
+      const startIndex = Math.max(0, viewportStartIndex - this.overScanCount);
+      const endIndex = Math.min(
+        this.totalRows - 1,
+        viewportEndIndex + this.overScanCount
+      );
+
+      const rows = this.getViewportRows(startIndex, endIndex);
+
+      this.stateMap.startIndex = startIndex;
+      this.stateMap.endIndex = endIndex;
+      this.stateMap.viewportRows = rows;
+    }
+  }
+
+  private layoutColumns = (): void => {
+    const visibleRowCount =
+      this.stateMap.endIndex - this.stateMap.startIndex + 1;
+    const diff = this.stateMap.startIndex - this.lastStartIndex;
+    this.lastStartIndex = this.stateMap.startIndex;
+
+    this.iterateColumns((col, pool) => {
+      pool.updateElements(visibleRowCount);
+
+      if (diff > 0) {
+        pool.swapHeadToTail(diff);
+      } else {
+        pool.swapTailToHead(-diff);
+      }
+
+      col.style.minHeight = `${this.rowHeight * this.totalRows}px`;
+
+      pool.iterateElements((el, binding, rowIndex) => {
+        const row = this.stateMap.viewportRows[rowIndex];
+        if (row != null) {
+          this.updateCell(
+            row,
+            el as HTMLVertexSceneTreeTableCellElement,
```

Comment:
I think we should assume that a user can swap out the cell to something other than a `vertex-scene-tree-table-cell`. This way we have an escape hatch if devs don't want to use our component.